### PR TITLE
Add `RetryingMetastore` for gRPC Metastore

### PR DESCRIPTION
**Pull request by boraarslan**

### CODE REVIEW

File: quickwit/quickwit-metastore/src/metastore/retrying_metastore/mod.rs [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+/// This is a band-aid solution for now. This will be removed after retry can be usable on
+/// tonic level.
+/// Tracking Issue: https://github.com/tower-rs/tower/issues/682
+pub struct RetryingMetastore {
+    inner: MetastoreGrpcClient,
+    retry_params: RetryParams,
+}
+
+impl RetryingMetastore {
+    /// Create a retry layer for [`MetastoreGrpcClient`]
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/metastore/retrying_metastore/mod.rs:45]:
```suggestion
    /// Creates a retry layer for [`MetastoreGrpcClient`]
```

File: quickwit/quickwit-metastore/src/metastore/retrying_metastore/mod.rs [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    pub fn new(metastore: MetastoreGrpcClient) -> Self {
+        Self {
+            inner: metastore,
+            retry_params: RetryParams {
+                max_attempts: 3,
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/metastore/retrying_metastore/mod.rs:50]:
```suggestion
                max_attempts: 5,
```

File: quickwit/quickwit-metastore/src/metastore/retrying_metastore/mod.rs [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        &self,
+        index_id: &str,
+        opstamp_start: u64,
+    ) -> MetastoreResult<Vec<DeleteTask>> {
+        retry(&self.retry_params, || async {
+            self.inner.list_delete_tasks(index_id, opstamp_start).await
+        })
+        .await
+    }
+}
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/metastore/retrying_metastore/mod.rs:264]: Let's add some tests.

File: quickwit/quickwit-metastore/src/metastore/retrying_metastore/mod.rs [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    IndexMetadata, ListSplitsQuery, Metastore, MetastoreGrpcClient, MetastoreResult, Split,
+    SplitMetadata,
+};
+
+/// Retry layer for [`MetastoreGrpcClient`].
+/// This is a band-aid solution for now. This will be removed after retry can be usable on
+/// tonic level.
+/// Tracking Issue: https://github.com/tower-rs/tower/issues/682
+pub struct RetryingMetastore {
+    inner: MetastoreGrpcClient,
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/metastore/retrying_metastore/mod.rs:40]: `dyn Metastore` will make testing easier.

File: quickwit/quickwit-metastore/src/checkpoint.rs [hunk 5]
```diff
@@ -266,7 +266,7 @@ impl<'de> Deserialize<'de> for SourceCheckpoint {
 /// Error returned when trying to apply a checkpoint delta to a checkpoint that is not
 /// compatible. ie: the checkpoint delta starts from a point anterior to
 /// the checkpoint.
-#[derive(Debug, Error, Eq, PartialEq, Serialize, Deserialize)]
+#[derive(Debug, Error, Eq, PartialEq, Serialize, Deserialize, Clone)]
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/checkpoint.rs:269]:
```suggestion
#[derive(Clone, Debug, Error, Eq, PartialEq, Serialize, Deserialize)]
```

File: quickwit/quickwit-metastore/src/error.rs [hunk 6]
```diff
@@ -25,7 +25,7 @@ use crate::checkpoint::IncompatibleCheckpointDelta;
 
 /// Metastore error kinds.
 #[allow(missing_docs)]
-#[derive(Debug, Error, Serialize, Deserialize, PartialEq, Eq)]
+#[derive(Debug, Error, Serialize, Deserialize, PartialEq, Eq, Clone)]
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/error.rs:28]:
```suggestion
#[derive(Clone, Debug, Error, Serialize, Deserialize, PartialEq, Eq)]
```

File: quickwit/quickwit-metastore/src/metastore/retrying_metastore/mod.rs [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+use self::retry::{retry, RetryParams};
+use crate::checkpoint::IndexCheckpointDelta;
+use crate::{IndexMetadata, ListSplitsQuery, Metastore, MetastoreResult, Split, SplitMetadata};
+
+/// Retry layer for [`MetastoreGrpcClient`].
+/// This is a band-aid solution for now. This will be removed after retry can be usable on
+/// tonic level.
+/// Tracking Issue: https://github.com/tower-rs/tower/issues/682
+pub struct RetryingMetastore {
+    inner: Arc<dyn Metastore>,
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/metastore/retrying_metastore/mod.rs:41]: `RetryingMetastore` is always going to be wrapped in an `Arc`, right? Can we get away with only `Box` here then?

File: quickwit/quickwit-metastore/src/metastore/retrying_metastore/test.rs [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+                source_type: "".to_string(),
+            },
+            MetastoreError::InternalError {
+                message: "".to_string(),
+                cause: "".to_string(),
+            },
+        ],
+    );
+
+    let result = metastore.list_indexes_metadatas().await;
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/metastore/retrying_metastore/test.rs:299]:
```rust
    let error = metastore.list_indexes_metadatas().await.unwrap_err();

    assert_eq!(
        error,
        MetastoreError::SourceAlreadyExists {
            source_id: "".to_string(),
            source_type: "".to_string(),
        }
    )
```

**Review by guilload — APPROVED** [review]
You went slightly overboard with the unit testing, but I'd take that any day over no tests at all.

**Comment by boraarslan** [comment]
> You went slightly overboard with the unit testing, but I'd take that any day over no tests at all.

I love tests

**Pull request merged**